intr_ptrace_decline test intermittently failing: trying to replay futex() but enter write()

This isn't a syscallbuf issue, in this failure at least, because it's happening with the syscallbuf disabled.

<pre>
48: Test 'intr_ptrace_decline' FAILED: error during replay:
48: --------------------------------------------------
48: time: 0
48: [ERROR] (/home/cjones/build/rr/src/share/util.c:644:compare_register_files: errno: None) ebx registers do not match: replaying: 55b62ba8 and recorded: 1
48: 
48: [ERROR] (/home/cjones/build/rr/src/share/util.c:652:compare_register_files: errno: None) ecx registers do not match: replaying: 0 and recorded: ffffd15c
48: 
48: [ERROR] (/home/cjones/build/rr/src/share/util.c:660:compare_register_files: errno: None) edx registers do not match: replaying: 4a05 and recorded: d
48: 
48: [ERROR] (/home/cjones/build/rr/src/share/util.c:676:compare_register_files: errno: None) edi registers do not match: replaying: 557b3ff4 and recorded: 0
48: 
48: [ERROR] (/home/cjones/build/rr/src/share/util.c:684:compare_register_files: errno: None) ebp registers do not match: replaying: 55b62b40 and recorded: ffffd568
48: 
48: [EMERGENCY] (/home/cjones/build/rr/src/share/util.c:735:assert_child_regs_are: errno: None) Assertion `regs_are_equal' failed to hold: [write in state 0, trace file line 568]
48: [INFO] (log_pending_events) (no pending events)
48: [FATAL] (/home/cjones/build/rr/src/replayer/replayer.c:1512:emergency_debug: errno: None) (stderr not a tty, aborting emergency debugging)
</pre>

--- a/src/recorder/rec_sched.c
+++ b/src/recorder/rec_sched.c
@@ -290,7 +290,7 @@ void rec_sched_deregister_thread(struct task** t_ptr)
 	sys_close(t->child_mem_fd);
 	close(t->desched_fd);
 
-	sys_ptrace_detach(tid);
+	detach_and_reap(t);
 
 	/* finally, free the memory */
 	sighandlers_unref(&t->sighandlers);

--- a/src/recorder/recorder.c
+++ b/src/recorder/recorder.c
@@ -189,6 +189,13 @@ static void handle_ptrace_event(struct task** tp)
 	}
 
 	case PTRACE_EVENT_EXIT:
+		if (EV_SYSCALL == t->ev->type
+		    && SYS_exit_group == t->ev->syscall.no
+		    && rec_sched_get_num_threads() > 1) {
+			log_warn("exit_group() with > 1 task; may misrecord CLONE_CHILD_CLEARTID memory race");
+			task_group_destabilize(t->task_group);
+		}
+
 		t->event = t->unstable ? USR_UNSTABLE_EXIT : USR_EXIT;
 		push_pseudosig(t,
 			       t->unstable ? EUSR_UNSTABLE_EXIT : EUSR_EXIT,
@@ -750,6 +757,7 @@ void record()
 		case EV_SIGNAL_DELIVERY: {
 			int unstable = signal_state_changed(t, by_waitpid);
 			if (unstable) {
+				log_warn("Delivered core-dumping signal; may misrecord CLONE_CHILD_CLEARTID memory race");
 				task_group_destabilize(t->task_group);
 			}
 			continue;

--- a/src/replayer/rep_sched.c
+++ b/src/replayer/rep_sched.c
@@ -133,8 +133,7 @@ void rep_sched_deregister_thread(struct task **t_ptr)
 	num_threads--;
 	assert(num_threads >= 0);
 
-	/* detatch the child process*/
-	sys_ptrace_detach(t->tid);
+	detach_and_reap(t);
 
 	sys_free((void**) t_ptr);
 }

--- a/src/replayer/replayer.c
+++ b/src/replayer/replayer.c
@@ -1348,6 +1348,22 @@ static void replay_one_trace_frame(struct dbg_context* dbg,
 		t->unstable = 1;
 		/* fall through */
 	case USR_EXIT:
+		/* If the task was killed by a terminating signal,
+		 * then it may have ended abruptly in a syscall or at
+		 * some other random execution point.  That's bad for
+		 * replay, because we detach from the task after we
+		 * replay its "exit".  Since we emulate signal
+		 * delivery, the task may happily carry on with
+		 * (non-emulated!) execution after we detach.  That
+		 * execution might include things like |rm -rf ~|.
+		 *
+		 * To ensure that the task really dies, we send it a
+		 * terminating signal here.  One would like to use
+		 * SIGKILL, but for not-understood reasons that causes
+		 * shutdown hangs when joining the exited tracee.
+		 * Other terminating signals have not been observed to
+		 * hang, so that's what's used here.. */
+		syscall(SYS_tkill, t->tid, SIGABRT);
 		rep_sched_deregister_thread(&t);
 		/* Early-return because |t| is gone now. */
 		return;

--- a/src/share/util.c
+++ b/src/share/util.c
@@ -22,6 +22,7 @@
 #include <sys/stat.h>
 #include <sys/syscall.h>
 #include <sys/types.h>
+#include <sys/wait.h>
 #include <time.h>
 #include <unistd.h>
 
@@ -208,6 +209,54 @@ int signal_pending(int status)
 		 * SEGVs? */
 		return sig & ~0x80;
 	}
+}
+
+void detach_and_reap(struct task* t)
+{
+	sys_ptrace_detach(t->tid);
+	if (t->unstable) {
+		log_warn("%d is unstable; not blocking on its termination",
+			 t->tid);
+		goto sleep_hack;
+	}
+
+	debug("Joining with exiting %d ...", t->tid);
+	while (1) {
+		int err = waitpid(t->tid, &t->status, __WALL);
+		if (-1 == err && ECHILD == errno) {
+			debug(" ... ECHILD");
+			break;
+		} else if (-1 == err) {
+			assert_exec(t, EINTR == errno,
+				    "waitpid(%d) returned -1, errno %d",
+				    t->tid, errno);
+		}
+		if (err == t->tid && (WIFEXITED(t->status) || 
+				      WIFSIGNALED(t->status))) {
+			debug(" ... exited with status 0x%x", t->status);
+			break;
+		} else if (err == t->tid) {
+			assert_exec(t, (PTRACE_EVENT_EXIT ==
+					GET_PTRACE_EVENT(t->status)),
+				    "waitpid(%d) return status %d",
+				    t->tid, t->status);
+		}
+	}
+
+sleep_hack:
+	/* clone()'d tasks can have a pid_t* |ctid| argument that's
+	 * written with the new task's pid.  That pointer can also be
+	 * used as a futex: when the task dies, the original ctid
+	 * value is cleared and a FUTEX_WAKE is done on the
+	 * address. So pthread_join() is basically a standard futex
+	 * wait loop.
+	 *
+	 * That means that the kernel writes shared memory behind rr's
+	 * back, which can diverge replay.  The "real fix" for this is
+	 * for rr to track access to shared memory, like the |ctid|
+	 * location.  But until then, we (attempt to) let "time"
+	 * resolve this memory race with the sleep() hack below. */
+	usleep(100);
 }
 
 void print_register_file_tid(pid_t tid)

--- a/src/share/util.h
+++ b/src/share/util.h
@@ -128,6 +128,11 @@ const char* syscallname(int syscall);
 
 int signal_pending(int status);
 
+/**
+ * Detach |t| from rr and try hard to ensure any operations related to
+ * it have completed by the time this function returns.
+ */
+void detach_and_reap(struct task* t);
 
 struct current_state_buffer {
 	pid_t pid;

--- a/src/test/term_nonmain.c
+++ b/src/test/term_nonmain.c
@@ -15,7 +15,7 @@ static void waittermsig(int sig, const char* waiter)
 	sigaddset(&set, sig);
 	sigtimedwait(&set, &si, &ts);
 
-	atomic_printf("FAILED: %s: signal %d either not caught or didn't terminate process within 1 second",
+	atomic_printf("FAILED: %s: signal %d either not caught or didn't terminate process within 1 second\n",
 		      waiter, sig);
 }
 

